[lldb] Add QSupported key to report watchpoint types supported

jasonmolenda opened the pull request:
debugserver on arm64 devices can manage both Byte Address Select watchpoints (1-8 bytes) and MASK watchpoints (8 bytes-2 gigabytes). This adds a SupportedWatchpointTypes key to the QSupported response from debugserver with a list of these, so lldb can take full advantage of them when creating larger regions with a single hardware watchpoint.

Also add documentation for this, and two other lldb extensions, to the lldb-gdb-remote.txt documentation.

Re-enable TestLargeWatchpoint.py on Darwin systems when testing with the in-tree built debugserver.  I can remove the "in-tree built debugserver" in the future when this new key is handled by an Xcode debugserver.

jasonmolenda committed:
[lldb] Add QSupported key to report watchpoint types supported

debugserver on arm64 devices can manage both Byte Address Select
watchpoints (1-8 bytes) and MASK watchpoints (8 bytes-2 gigabytes).
This adds a SupportedWatchpointTypes key to the QSupported response
from debugserver with a list of these, so lldb can take full advantage
of them when creating larger regions with a single hardware watchpoint.

Also add documentation for this, and two other lldb extensions, to
the lldb-gdb-remote.txt documentation.

Re-enable TestLargeWatchpoint.py on Darwin systems when testing with
the in-tree built debugserver.  I can remove the "in-tree built
debugserver" in the future when this new key is handled by an Xcode
debugserver.

jasonmolenda commented:
The QSupported packet from debugserver also has a "compression minimum size" field it adds (see the change to RNBRemote.cpp) and the QEnableCompression packet accepts that field as well.  I must have had a reason for adding this when I added compression to the protocol, but lldb doesn't use this.  I'll remove it in a separate patch.

jasonmolenda committed:
Clarify 'size of pointer' language.

DavidSpickett commented on lldb/docs/lldb-gdb-remote.txt:
```diff
@@ -38,7 +38,44 @@ read packet: +
+
+read packet: qXfer:features:read+;PacketSize=20000;qEcho+;native-signals+;SupportedCompressions=lzfse,zlib-deflate,lz4,lzma;SupportedWatchpointTypes =aarch64-mask,aarch64-bas;
+
+In this example, three lldb extensions are reported:
```
Is the total number of lldb extensions 3? Might as well say "the example shows all the extensions".

If you don't know (I don't either) how many extensions we have, let's imagine I never left this comment :)

jasonmolenda replied:
I don't have high confidence that the documentation will be updated as people add new extensions, so I was reluctant to make it sound like the extensions were exhaustive.

DavidSpickett commented on lldb/docs/lldb-gdb-remote.txt:
```diff
@@ -38,7 +38,44 @@ read packet: +
+
+send packet: qSupported:xmlRegisters=i386,arm,mips,arc;multiprocess+;fork-events+;vfork-events+
+
+read packet: qXfer:features:read+;PacketSize=20000;qEcho+;native-signals+;SupportedCompressions=lzfse,zlib-deflate,lz4,lzma;SupportedWatchpointTypes =aarch64-mask,aarch64-bas;
```
Is there really a space in `SupportedWatchpointTypes =aarch64-mask` or just a typo?

jasonmolenda replied:
lol, copy & paste had spread it across two lines.  I joined them but missed the space that got introduced.

DavidSpickett commented on lldb/docs/lldb-gdb-remote.txt:
```diff
@@ -38,7 +38,44 @@ read packet: +
+
+In this example, three lldb extensions are reported:
+  PacketSize=20000
+    The base16 maximum packet size that the GDB Remote Serial stub
```
"base 16"?

Also what is a serial stub? Does that literally mean when we use a serial port to talk to the debug server?

jasonmolenda replied:
I could say "the hex maximum packet size", I suppose, is that clearer?  One of the annoying things (IMO) about gdb remote serial protocol is that many numbers are hex numbers without a 0x prefix, but sometimes (probably mistakes) they're base 10.  So if you see a number like "2000" in a log file, you don't know what that number is.

The naming in this text file is a bit all over the place.  The protocol is called "gdb remote serial protocol".  The opposite side from lldb, debugserver or lldb-server, is called the stub.  Some places call it the "gdb RSP stub", some just say "stub".

jasonmolenda replied:
(incidentally debugserver also reports a compression minimum size -- which I'll be removing after this patch is landed because it does nothing -- and that number is sent as a base 10 number.  lol it's a wild west for numbers.)

DavidSpickett commented on lldb/docs/lldb-gdb-remote.txt:
```diff
@@ -38,7 +38,44 @@ read packet: +
+                     (any power-of-2 region of memory from 8 to 2GB, aligned)
+
+lldb will default to sending power-of-2 watchpoints up to a pointer size,
+`sizeof(void*)`, in the target process if nothing is specified.
```
I would rephrase and put "If nothing is specified" first, so the reader knows immediately what the point of this is.

Also tab it in to be part of the `SupportedWatchpointTypes` block.

DavidSpickett commented on lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationClient.cpp:
```diff
@@ -403,6 +403,22 @@ void GDBRemoteCommunicationClient::GetRemoteQSupported() {
+          if (wp_type == "aarch64-bas")
+            m_watchpoint_types |=
+                WatchpointHardwareFeature::eWatchpointHardwareArmBAS;
+        }
```
Please put this string -> enum code next to the enum if possible. Perhaps `eWatchpointHardwareFeatureUnknown` could be the fallback value, assuming that that is 0.

Then you can loop over the strings here doing:
m_watchpoint_types |= stringToType(..);

And we'll have one place to update/reference in future.

DavidSpickett replied:
Also I wonder if this should be case insensitive, but I doubt it'll ever be an issue. If compression types does that then sure, if not, it's fine as is.

jasonmolenda replied:
I thought about putting the gdb remote serial protocol names in lldb-enumerations.h but they're gdb RSP specific strings and I don't expect them to be used anywhere else.

DavidSpickett replied:
Sure. The other motivation for me was to not add more else if to `GetRemoteQSupported` which is probably a gazillion lines long. But a few more on top of that isn't changing anything so it's fine.

DavidSpickett commented:
Just to record that I thought about it, I agree with not trying to make the names architecture neutral.

For example one might say that because range watchpoints on mips and aarch64 have overlapping functionality (no pun intended) that we could report that we have "range" as a type for either of them. This leads to a situation where the functionality that doesn't overlap becomes much harder to handle and detect.

So I agree with `aarch64-mask` instead of simply `mask`, as an example.

If we later want to (and kinda already do) group these by general functionality, we can do that in lldb in the algorithms you added for working out the watchpoint resources.

(and sod's law, someone will propose to add `aarch64-bas-but-this-one-thing-is-different` to the architecture one of these days :) )

jasonmolenda commented:
> Just to record that I thought about it, I agree with not trying to make the names architecture neutral.
> 
> For example one might say that because range watchpoints on mips and aarch64 have overlapping functionality (no pun intended) that we could report that we have "range" as a type for either of them. This leads to a situation where the functionality that doesn't overlap becomes much harder to handle and detect.
> 
> So I agree with `aarch64-mask` instead of simply `mask`, as an example.
> 
> If we later want to (and kinda already do) group these by general functionality, we can do that in lldb in the algorithms you added for working out the watchpoint resources.
> 
> (and sod's law, someone will propose to add `aarch64-bas-but-this-one-thing-is-different` to the architecture one of these days :) )

Yeah I thought about specifying a more structured name, like "ARCH-STYLE" so we'd have "x86_64-mask" in addition to "aarch64-mask", but my guess is a lot of targets will work fine with "power of 2, up to sizeof(void*)" and never need to adjust this.  Ted Woodward pointed out that the PowerPC watchpoints are quite different, so if we support that target we'll definitely need an additional algorithm for sure.

jasonmolenda committed:
Address David's initial feedback comments

jasonmolenda committed:
Word tweak in doc.

jasonmolenda committed:
Move the WatchpointHardwareFeature to lldb-private-enumerations

WatchpointHardwareFeature is set by the Process plugin to
WatchpointAlgorithms::AtomizeWatchpointRequest to indicate which
types of watchpoints this target/stub support.  It's entirely
internal and should be in lldb-private-enumerations.h.

To make a bitfield enum work with the bitfield, & and |, operators
without lots of casting, I need the LLDB_MARK_AS_BITMASK_ENUM()
constexpr template stuff from lldb-enumerations.h. It might be
better to put this and FLAGS_ENUM() in their own file, but I
don't think I'm messing up with any layering violations by having
lldb-private-enumerations.h include lldb-enumerations.h to get
them so I'll start with this.

jasonmolenda commented:
@bulbazord in the most recent commit I moved this internal-only enum from lldb-enumerations.h to lldb-private-enumerations.h, but I need to use the constexpr templatey thing that LLDB_MARK_AS_BITMASK_ENUM() defines for the enum so I can use bit-wise | & operations without casting everywhere; that's defined in lldb-enumerations.h so I included the public enums in the lldb-private-enumerations.h.  It seems like it's probably not a great choice, but the other one is breaking out this and FLAGS_ENUM etc into a little lldb-common-enumerations.h or something.  What do you think?

DavidSpickett commented on lldb/docs/lldb-gdb-remote.txt:
```diff
@@ -38,7 +38,43 @@ read packet: +
+In this example, three lldb extensions are shown:
+
+  PacketSize=20000
+    The base16 maximum packet size that the stub can handle.
```
I was mostly confused by the lack of a space between base and 16. "base 16" is recognisable, "base16" sounds like a nightclub.I was mostly confused by the lack of a space between base and 16. "base 16" is recognisable, "base16" sounds like a nightclub.

jasonmolenda replied:
OK, I can see that.  I fixed this and a handful of other basenn uses in the file (probably all from me over the years).

DavidSpickett commented on lldb/docs/lldb-gdb-remote.txt:
```diff
@@ -38,7 +38,43 @@ read packet: +
+
+read packet: qXfer:features:read+;PacketSize=20000;qEcho+;native-signals+;SupportedCompressions=lzfse,zlib-deflate,lz4,lzma;SupportedWatchpointTypes=aarch64-mask,aarch64-bas;
+
+In this example, three lldb extensions are shown:
```
"In the example above"

DavidSpickett commented on lldb/docs/lldb-gdb-remote.txt:
```diff
@@ -38,7 +38,43 @@ read packet: +
+//----------------------------------------------------------------------
 
+QSupported is a standard GDB Remote Serial Protocol packet, but
+there are several additions to the response that lldb can parse.
```
"can parse. They are not all listed here.
<blank line>
An example exchange:"

bulbazord commented:
> @bulbazord in the most recent commit I moved this internal-only enum from lldb-enumerations.h to lldb-private-enumerations.h, but I need to use the constexpr templatey thing that LLDB_MARK_AS_BITMASK_ENUM() defines for the enum so I can use bit-wise | & operations without casting everywhere; that's defined in lldb-enumerations.h so I included the public enums in the lldb-private-enumerations.h. It seems like it's probably not a great choice, but the other one is breaking out this and FLAGS_ENUM etc into a little lldb-common-enumerations.h or something. What do you think?

Not the greatest thing in the world but it's not terrible either. We use the lldb public enumerations everywhere in private code, so as long as we're not going the other way this is ok to do. I will say, by moving the definitions from `lldb-enumerations.h` to `lldb-private-enumerations.h` you have removed these values from the python bindings. Technically that's an API break, but I'm not sure where anyone could have used these values otherwise. LGTM.

jasonmolenda committed:
Space between "base" and the number, consistently.
Addressing feedback from David Spickett.

jasonmolenda committed:
Two more copyedit suggestions from David in the docs.

jasonmolenda commented:
Thanks for the feedback Alex & David, I think this one might be good enough now.

JDevlieghere reviewed (approved):
LGTM if @DavidSpickett and @bulbazord are happy.

bulbazord reviewed (approved):
LGTM
